File: rust-src/data/jsonl_parser.rs
//! JSONL parser for Claude Code session logs.

use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use serde_json::Value;

use crate::models::{TokenUsage, UsageRecord};


/// Parse a single JSONL file and return UsageRecord objects.
pub fn parse_jsonl_file(file_path: &Path) -> Result<Vec<UsageRecord>> {
    let file = File::open(file_path)
        .with_context(|| format!("Failed to open file: {}", file_path.display()))?;

    let reader = BufReader::new(file);
    let mut records = Vec::new();

    for (line_num, line_result) in reader.lines().enumerate() {
        let line = match line_result {
            Ok(l) => l,
            Err(e) => {
                eprintln!(
                    "Warning: Error reading line {} in {}: {}",
                    line_num + 1,
                    file_path.display(),
                    e
                );
                continue;
            }
        };

        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        match serde_json::from_str::<Value>(line) {
            Ok(data) => {
                if let Some(record) = parse_record(&data) {
                    records.push(record);
                }
            }
            Err(e) => {
                eprintln!(
                    "Warning: Skipping malformed JSON at {}:{}: {}",
                    file_path.display(),
                    line_num + 1,
                    e
                );
            }
        }
    }

    Ok(records)
}


/// Parse multiple JSONL files and return all usage records.
pub fn parse_all_jsonl_files(file_paths: &[&Path]) -> Result<Vec<UsageRecord>> {
    if file_paths.is_empty() {
        anyhow::bail!("No JSONL files provided to parse");
    }

    let mut all_records = Vec::new();

    for file_path in file_paths {
        match parse_jsonl_file(file_path) {
            Ok(records) => all_records.extend(records),
            Err(e) => {
                eprintln!("Warning: Error parsing {}: {}", file_path.display(), e);
            }
        }
    }

    Ok(all_records)
}


/// Parse a single JSON record into a UsageRecord.
fn parse_record(data: &Value) -> Option<UsageRecord> {
    let message_type = data.get("type")?.as_str()?;

    // Only process user and assistant messages
    if message_type != "user" && message_type != "assistant" {
        return None;
    }

    // Parse timestamp
    let timestamp_str = data.get("timestamp")?.as_str()?;
    let timestamp = parse_timestamp(timestamp_str)?;

    // Extract metadata
    let session_id = data
        .get("sessionId")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();

    let message_uuid = data
        .get("uuid")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();

    let folder = data
        .get("cwd")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();

    let git_branch = data
        .get("gitBranch")
        .and_then(|v| v.as_str())
        .map(String::from);

    let version = data
        .get("version")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();

    // Extract message data
    let message = data.get("message")?;
    let model = message.get("model").and_then(|v| v.as_str()).map(String::from);

    // Filter out synthetic models
    if model.as_deref() == Some("<synthetic>") {
        return None;
    }

    // Extract content for analysis
    let (content, char_count) = extract_content(message);

    // Extract token usage (only for assistant messages)
    let token_usage = if message_type == "assistant" {
        extract_token_usage(message)
    } else {
        None
    };

    Some(UsageRecord {
        timestamp,
        session_id,
        message_uuid,
        message_type: message_type.to_string(),
        model,
        folder,
        git_branch,
        version,
        token_usage,
        content,
        char_count,
    })
}


/// Parse ISO 8601 timestamp string to DateTime<Utc>.
fn parse_timestamp(s: &str) -> Option<DateTime<Utc>> {
    // Handle "Z" suffix
    let normalized = s.replace("Z", "+00:00");
    DateTime::parse_from_rfc3339(&normalized)
        .ok()
        .map(|dt| dt.with_timezone(&Utc))
}


/// Extract content and character count from message.
fn extract_content(message: &Value) -> (Option<String>, i64) {
    let content_val = message.get("content");

    match content_val {
        Some(Value::String(s)) => {
            let char_count = s.len() as i64;
            (Some(s.clone()), char_count)
        }
        Some(Value::Array(blocks)) => {
            let mut text_parts = Vec::new();
            for block in blocks {
                if let Some(block_obj) = block.as_object() {
                    if block_obj.get("type").and_then(|t| t.as_str()) == Some("text") {
                        if let Some(text) = block_obj.get("text").and_then(|t| t.as_str()) {
                            text_parts.push(text.to_string());
                        }
                    }
                }
            }
            if text_parts.is_empty() {
                (None, 0)
            } else {
                let content = text_parts.join("\n");
                let char_count = content.len() as i64;
                (Some(content), char_count)
            }
        }
        _ => (None, 0),
    }
}


/// Extract token usage from assistant message.
fn extract_token_usage(message: &Value) -> Option<TokenUsage> {
    let usage = message.get("usage")?;

    let input_tokens = usage
        .get("input_tokens")
        .and_then(|v| v.as_i64())
        .unwrap_or(0);

    let output_tokens = usage
        .get("output_tokens")
        .and_then(|v| v.as_i64())
        .unwrap_or(0);

    // Cache creation tokens (sum of all cache creation fields)
    let cache_creation = usage.get("cache_creation");
    let cache_creation_tokens = if let Some(cc) = cache_creation {
        let base = cc
            .get("cache_creation_input_tokens")
            .and_then(|v| v.as_i64())
            .unwrap_or(0);
        let ephemeral_5m = cc
            .get("ephemeral_5m_input_tokens")
            .and_then(|v| v.as_i64())
            .unwrap_or(0);
        let ephemeral_1h = cc
            .get("ephemeral_1h_input_tokens")
            .and_then(|v| v.as_i64())
            .unwrap_or(0);
        base + ephemeral_5m + ephemeral_1h
    } else {
        0
    };

    let cache_read_tokens = usage
        .get("cache_read_input_tokens")
        .and_then(|v| v.as_i64())
        .unwrap_or(0);

    Some(TokenUsage {
        input_tokens,
        output_tokens,
        cache_creation_tokens,
        cache_read_tokens,
    })
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_timestamp() {
        let ts = parse_timestamp("2024-01-15T10:30:00Z").unwrap();
        assert_eq!(ts.to_rfc3339(), "2024-01-15T10:30:00+00:00");
    }

    #[test]
    fn test_parse_record_user() {
        let json = r#"{
            "type": "user",
            "timestamp": "2024-01-15T10:30:00Z",
            "sessionId": "sess-123",
            "uuid": "msg-456",
            "cwd": "/home/user/project",
            "version": "1.0.0",
            "message": {
                "content": "Hello world"
            }
        }"#;

        let data: Value = serde_json::from_str(json).unwrap();
        let record = parse_record(&data).unwrap();

        assert_eq!(record.message_type, "user");
        assert_eq!(record.session_id, "sess-123");
        assert_eq!(record.content, Some("Hello world".to_string()));
        assert!(record.token_usage.is_none());
    }

    #[test]
    fn test_parse_record_assistant_with_usage() {
        let json = r#"{
            "type": "assistant",
            "timestamp": "2024-01-15T10:30:00Z",
            "sessionId": "sess-123",
            "uuid": "msg-789",
            "cwd": "/home/user/project",
            "version": "1.0.0",
            "message": {
                "model": "claude-sonnet-4-20250514",
                "content": "Here's the answer",
                "usage": {
                    "input_tokens": 100,
                    "output_tokens": 50,
                    "cache_read_input_tokens": 25
                }
            }
        }"#;

        let data: Value = serde_json::from_str(json).unwrap();
        let record = parse_record(&data).unwrap();

        assert_eq!(record.message_type, "assistant");
        assert_eq!(record.model, Some("claude-sonnet-4-20250514".to_string()));

        let usage = record.token_usage.unwrap();
        assert_eq!(usage.input_tokens, 100);
        assert_eq!(usage.output_tokens, 50);
        assert_eq!(usage.cache_read_tokens, 25);
    }

    #[test]
    fn test_skip_synthetic_model() {
        let json = r#"{
            "type": "assistant",
            "timestamp": "2024-01-15T10:30:00Z",
            "sessionId": "sess-123",
            "uuid": "msg-789",
            "cwd": "/home/user/project",
            "version": "1.0.0",
            "message": {
                "model": "<synthetic>",
                "content": "test"
            }
        }"#;

        let data: Value = serde_json::from_str(json).unwrap();
        assert!(parse_record(&data).is_none());
    }

    #[test]
    fn test_content_blocks() {
        let json = r#"{
            "type": "user",
            "timestamp": "2024-01-15T10:30:00Z",
            "sessionId": "sess-123",
            "uuid": "msg-456",
            "cwd": "/project",
            "version": "1.0.0",
            "message": {
                "content": [
                    {"type": "text", "text": "Hello"},
                    {"type": "text", "text": "World"}
                ]
            }
        }"#;

        let data: Value = serde_json::from_str(json).unwrap();
        let record = parse_record(&data).unwrap();

        assert_eq!(record.content, Some("Hello\nWorld".to_string()));
        assert_eq!(record.char_count, 11);
    }
}

File: src/data/jsonl_parser.py
#region Imports
import json
from collections.abc import Iterator
from datetime import datetime
from pathlib import Path

from src.models.usage_record import TokenUsage, UsageRecord

#endregion


#region Functions


def parse_jsonl_file(file_path: Path) -> Iterator[UsageRecord]:
    """
    Parse a single JSONL file and yield UsageRecord objects.

    Extracts usage data from Claude Code session logs, including:
    - Token usage (input, output, cache creation, cache read)
    - Session metadata (model, folder, version, branch)
    - Timestamps and identifiers

    Args:
        file_path: Path to the JSONL file to parse

    Yields:
        UsageRecord objects for each assistant message with usage data

    Raises:
        FileNotFoundError: If the file doesn't exist
        json.JSONDecodeError: If the file contains invalid JSON
    """
    if not file_path.exists():
        raise FileNotFoundError(f"File not found: {file_path}")

    with open(file_path, encoding="utf-8") as f:
        for line_num, line in enumerate(f, start=1):
            line = line.strip()
            if not line:
                continue

            try:
                data = json.loads(line)
                record = _parse_record(data)
                if record:
                    yield record
            except json.JSONDecodeError as e:
                # Skip malformed lines but continue processing
                print(f"Warning: Skipping malformed JSON at {file_path}:{line_num}: {e}")
                continue


def parse_all_jsonl_files(file_paths: list[Path]) -> list[UsageRecord]:
    """
    Parse multiple JSONL files and return deduplicated usage records.

    Assistant records are deduplicated GLOBALLY by their billed-response
    identity (message_uuid = API message id + request id): Claude Code
    writes one entry per streaming flush and session forks replay history
    into new session files, so one billed response can appear many times
    across entries, sessions, and files. The record with the largest total
    token usage wins (usage grows monotonically across flushes); ties keep
    the latest timestamp. User records pass through untouched.

    Args:
        file_paths: List of paths to JSONL files

    Returns:
        List of deduplicated UsageRecord objects across all files

    Raises:
        ValueError: If file_paths is empty
    """
    if not file_paths:
        raise ValueError("No JSONL files provided to parse")

    records: list[UsageRecord] = []
    for file_path in file_paths:
        try:
            records.extend(parse_jsonl_file(file_path))
        except FileNotFoundError:
            print(f"Warning: File not found, skipping: {file_path}")
        except Exception as e:
            print(f"Warning: Error parsing {file_path}: {e}")

    return dedupe_records(records)


def dedupe_records(records: list[UsageRecord]) -> list[UsageRecord]:
    """
    Collapse assistant records sharing a billed-response identity.

    Keeps, per message_uuid, the record with max total tokens (ties: latest
    timestamp), preserving first-appearance order. User records pass through.
    """
    best: dict[str, UsageRecord] = {}
    order: list[str] = []
    others: list[UsageRecord] = []
    for record in records:
        if not record.is_assistant_response:
            others.append(record)
            continue
        key = record.message_uuid
        current = best.get(key)
        if current is None:
            best[key] = record
            order.append(key)
            continue
        new_total = record.token_usage.total_tokens if record.token_usage else 0
        cur_total = current.token_usage.total_tokens if current.token_usage else 0
        if new_total > cur_total or (
            new_total == cur_total and record.timestamp > current.timestamp
        ):
            best[key] = record
    return others + [best[key] for key in order]


def _parse_record(data: dict) -> UsageRecord | None:
    """
    Parse a single JSON record into a UsageRecord.

    Processes both user prompts and assistant responses.
    Skips system events and other message types.

    Args:
        data: Parsed JSON object from JSONL line

    Returns:
        UsageRecord for user or assistant messages, None otherwise
    """
    message_type = data.get("type")

    # Only process user and assistant messages
    if message_type not in ("user", "assistant"):
        return None

    # Parse timestamp
    timestamp_str = data.get("timestamp")
    if not timestamp_str:
        return None

    timestamp = datetime.fromisoformat(timestamp_str.replace("Z", "+00:00"))

    # Extract metadata (common to both user and assistant)
    session_id = data.get("sessionId", "unknown")
    folder = data.get("cwd", "unknown")
    git_branch = data.get("gitBranch")
    version = data.get("version", "unknown")

    # Extract message data
    message = data.get("message", {})
    model = message.get("model")

    # Identity: assistant rows key on the billed API response (message id +
    # request id) so streaming flush entries and session-fork replays of the
    # same response dedupe to one record; user/legacy rows keep the
    # transcript entry uuid.
    api_id = message.get("id")
    request_id = data.get("requestId")
    if message_type == "assistant" and api_id:
        message_uuid = f"{api_id}:{request_id}" if request_id else api_id
    else:
        message_uuid = data.get("uuid", "unknown")

    # Filter out synthetic models (test/internal artifacts)
    if model == "<synthetic>":
        return None

    # Extract content for analysis
    content = None
    char_count = 0
    if isinstance(message.get("content"), str):
        content = message["content"]
        char_count = len(content)
    elif isinstance(message.get("content"), list):
        # Handle content blocks (concatenate text)
        text_parts = []
        for block in message["content"]:
            if isinstance(block, dict) and block.get("type") == "text":
                text_parts.append(block.get("text", ""))
        content = "\n".join(text_parts) if text_parts else None
        char_count = len(content) if content else 0

    # Extract token usage (only available for assistant messages)
    token_usage = None
    if message_type == "assistant":
        usage_data = message.get("usage")
        if usage_data:
            cache_creation = usage_data.get("cache_creation", {})
            cache_creation_tokens = (
                cache_creation.get("cache_creation_input_tokens", 0)
                + cache_creation.get("ephemeral_5m_input_tokens", 0)
                + cache_creation.get("ephemeral_1h_input_tokens", 0)
            )

            token_usage = TokenUsage(
                input_tokens=usage_data.get("input_tokens", 0),
                output_tokens=usage_data.get("output_tokens", 0),
                cache_creation_tokens=cache_creation_tokens,
                cache_read_tokens=usage_data.get("cache_read_input_tokens", 0),
                cache_creation_1h_tokens=cache_creation.get("ephemeral_1h_input_tokens", 0),
            )

    return UsageRecord(
        timestamp=timestamp,
        session_id=session_id,
        message_uuid=message_uuid,
        message_type=message_type,
        model=model,
        folder=folder,
        git_branch=git_branch,
        version=version,
        token_usage=token_usage,
        content=content,
        char_count=char_count,
    )
#endregion
